test/Test.agda

```diff
@@ -1,76 +1,144 @@
-
 module _ where
 
 open import Agda.Builtin.List
 open import Agda.Builtin.Nat
+-- open import Agda.Builtin.Float
+-- open import Agda.Builtin.Word
+-- open import Agda.Builtin.Char
 open import Agda.Builtin.Equality
 
 variable
   a b : Set
 
+-- ** Foreign HS code
+
+-- language extensions
 {-# FOREIGN AGDA2HS
 {-# LANGUAGE LambdaCase #-}
 #-}
 
+-- imports
 {-# FOREIGN AGDA2HS
 import Prelude hiding (map, sum, (++))
 import Data.Monoid
+import Data.Word
+#-}
+
+-- comments
+{-# FOREIGN AGDA2HS
+-- comment
+-- another comment
+bla :: Int -> Int
+bla n = n * 4
+
+{- multi
+   line
+   comment
+-}
 #-}
 
+-- ** Datatypes & functions
+
 data Exp (v : Set) : Set where
   Plus : Exp v → Exp v → Exp v
   Int : Nat → Exp v
   Var : v → Exp v
-
 {-# COMPILE AGDA2HS Exp #-}
 
 eval : (a → Nat) → Exp a → Nat
 eval env (Plus a b) = eval env a + eval env b
 eval env (Int n) = n
 eval env (Var x) = env x
-
 {-# COMPILE AGDA2HS eval #-}
 
+-- ** Natural numbers
+
 sum : List Nat → Nat
 sum []       = 0
 sum (x ∷ xs) = x + sum xs
-
 {-# COMPILE AGDA2HS sum #-}
 
-{-# FOREIGN AGDA2HS
--- comment
--- another comment
-bla :: Int -> Int
-bla n = n * 4
+-- ** Extra builtins
 
-{- multi
-   line
-   comment
--}
-#-}
+-- floats
+postulate Float : Set
+{-# BUILTIN FLOAT Float #-}
+
+ex_float : Float
+ex_float = 0.0
+{-# COMPILE AGDA2HS ex_float #-}
+
+primitive
+  primFloatPlus : Float -> Float -> Float
+  -- TODO more primitives
+{-# COMPILE AGDA2HS primFloatPlus #-}
+
+sumF : List Float → Float
+sumF []       = 0.0
+sumF (x ∷ xs) = primFloatPlus x (sumF xs)
+{-# COMPILE AGDA2HS sumF #-}
+
+-- words
+postulate Word64 : Set
+{-# BUILTIN WORD64 Word64 #-}
+
+primitive
+  primWord64FromNat : Nat → Word64
+  primWord64ToNat : Word64 → Nat
+{-# COMPILE AGDA2HS primWord64FromNat = fromIntegral #-}
+{-# COMPILE AGDA2HS primWord64ToNat #-}
+
+ex_word : Word64
+ex_word = primWord64FromNat 0
+{-# COMPILE AGDA2HS ex_word #-}
+
+ex_nat : Nat
+ex_nat = primWord64ToNat ex_word
+{-# COMPILE AGDA2HS ex_nat #-}
+
+-- chars
+postulate Char : Set
+{-# BUILTIN CHAR Char #-}
+
+primitive
+  primCharToNat : Char → Nat
+  primNatToChar : Nat → Char
+  -- TODO more primitives
+{-# COMPILE AGDA2HS primCharToNat #-}
+{-# COMPILE AGDA2HS primNatToChar #-}
```

Comment:
This isn't really how you want this to work. Anything which doesn't let you simply import `Agda.Builtin.*` is a no-go in my opinion.

Reply:
Hmm, how would you go about this though? If there is no user PRAGMA to instruct `agda2hs` to compile a primitive, would we internally keep a (fixed) set of primitives and replace all occurrences in terms?

Reply:
Or something like an RTS.hs, which is imported by default and includes the primitive definitions in Haskell?

Reply:
I think we really want to avoid any `RTS.hs` things here. The idea is that you are using Agda to write (and prove things about) a Haskell program, so we'd like the end result to be a Haskell program that you might have written directly in Haskell. Having a special RTS module kind of defeats that.

Given this premise I think this can be handled as a library on the Agda side. The question isn't "How should we translate `primNatToChar` (say)?". The question is "How do we represent `fromEnum :: Char -> Int` in Agda?". 

It's perfectly fine if there are lots of Agda primitives/constructions that we don't translate.

Reply:
How about we postulate a marker that interprets a given HS string to any type and then the backend emits the given string immediately?
```agda
-- ** Extra builtins

postulate
  $ : ∀ {A : Set} → String → A

-- floats
ex_float : Float
ex_float = 0.0
{-# COMPILE AGDA2HS ex_float #-}

sumF : List Float → Float
sumF []       = 0.0
sumF (x ∷ xs) = $ {_ → _ → _} "(+) :: Double -> Double -> Double" x (sumF xs)
{-# COMPILE AGDA2HS sumF #-}

-- chars
ex_char : Char
ex_char = 'x'
{-# COMPILE AGDA2HS ex_char #-}

ex_toℕ : Nat
ex_toℕ = $ {_ → _} "(fromIntegral . fromEnum :: Char -> Integer)" 'd'
{-# COMPILE AGDA2HS ex_toℕ #-}

baseNat : Nat
baseNat = $ {Char → _} "(fromIntegral . fromEnum :: Char -> Integer)" ($ "toEnum 0")
{-# COMPILE AGDA2HS baseNat #-}
```

Reply:
This looks terrible. The reason to have the code in Agda is that you want to prove something about it. Having a postulate defeats that completely.

What is that actual problem that you want to solve here? I'm not convinced we need to go any further than #12.

Reply:
Yeah, you're right. Let's archive this PR.